### PR TITLE
Add strings to VM

### LINKED ISSUE #36: Add strings to VM

### PR DESCRIPTION
**Pull request by bwdmonkey**
:tickets: **Ticket(s)**: Closes #36 

---

## :construction_worker: Changes

- Added string type to DurianObject
- Implemented `NEWSTR8` opcode
- Placeholdered `CNCT` opcode
- Reformatted `opcode.h` (enum struct was indented with 2 spaces. Changed to soft-tab)

## :flashlight: Testing Instructions

Compiles? && Test out the error logging

### CODE REVIEW

File: src/include/DurianObject.h [hunk 1]
```diff
@@ -2,30 +2,29 @@
...
+    Boolean,
+    String
 };
 
 struct DurianObject {
     DurianObject();
     DurianObject(double dval);
     DurianObject(int64_t ival);
     DurianObject(bool bval);
+    DurianObject(int32_t len, unsigned char *sval);
```
Comment (bobheadxi) [inline src/include/DurianObject.h:19]: `sval`, `ival`, etc seems like an awkward naming scheme hmm

File: src/include/opcode.h [hunk 2]
```diff
@@ -3,26 +3,27 @@
...
+        ADD      = 0x60, // add top two elements on stack, push result back onto stack.
+        SUB      = 0x61, // subtract top two elements on stack, push result back onto stack.
+        MUL      = 0x62, // multiply top two elements on stack, push result back onto stack.
+        FDIV     = 0x64, // divide top two elements on stack, push result back onto stack as a double.
+        NEG      = 0x67, // negate top element on stack, push result back onto stack.
+        NOT      = 0x69, // negates a boolean on top of the stack.
+        CNCT     = 0x6B, // concatenate top two elements on stack, push result back onto stack.
+        EQ       = 0x74, // checks if the top two elements of the stack are equal.
+        BR_F     = 0x93, // pops the top of the stack, then branches if it is false. (takes next 4 bytes as jump length)
+        NEWSTR8  = 0xBB, // make new String (length: 8 bytes, value: length bytes) and push it onto stack
```
Comment (bobheadxi) [inline src/include/opcode.h:24]: note to self - this is new

File: src/interpreter/main.cpp [hunk 3]
```diff
@@ -6,7 +6,7 @@
 using namespace Opcode;
 
 int main() {
-    unsigned char bytecodes[] = {BCONST_F, DUP, BR_F, 0x02, 0x00, 0x00, 0x00, POP, BCONST_T, PRINT, HALT};
+    unsigned char bytecodes[] = {HALT};
```
Comment (bobheadxi) [inline src/interpreter/main.cpp]: maybe add a note that this is temporary (I presume?)

File: src/interpreter/vm/VM.cpp [hunk 4]
```diff
@@ -17,21 +17,24 @@ VM::VM(unsigned char *bytecode) :
...
 
 int VM::run() {
-    //DurianObject a, b, v;
     while(true) {
         unsigned char opcode = nextBytecode();
         DURIAN_DEBUG_LOG("%x\n", opcode);
         DurianObject a, b;
-        int32_t jumpLen;
+        int32_t jump_len; // Jump length
+        unsigned char *p_str; // String (length: 8 bytes, value: length bytes) pointer
```
Comment (bobheadxi) [inline src/interpreter/vm/VM.cpp]: why not name this something like `p_new_str` or `p_assigned_str`?

File: src/interpreter/vm/VM.cpp [hunk 5]
```diff
@@ -165,12 +168,25 @@ int VM::run() {
                     exit(EXIT_FAILURE);
                 }
                 break;
+            case Opcode::CNCT:
+                b = pop();
+                a = pop();
+                if (a.type == DurianType::String && b.type == DurianType::String) {
+                    push(DurianObject(a.value.sval.len /* + b.value.sval.len */, a.value.sval.val /* TODO: Implement concat string */));
+                } else {
+                    std::cout << "TypeError: Invalid operand types for &: "
```
Comment (bobheadxi) [inline src/interpreter/vm/VM.cpp]: can we have a helper function for this?

```cpp
    // ...
    ParserErrors::ReportTypeError( /* ... invalid operands or something */ )
    // ...
```

so we can get some standardized printing and management, will make it easier to flesh things out later instead of leaving `std::cout` everywhere

**Review by bobheadxi — COMMENTED** [review]
good work lad

File: src/interpreter/main.cpp (hunk 3 above)
Comment (rwblickhan) [reply src/interpreter/main.cpp]: Agreed.

File: src/interpreter/vm/DurianObject.cpp [hunk 6]
```diff
@@ -22,6 +22,12 @@ DurianObject::DurianObject(bool bval) :
     value.bval = bval;
 }
 
+DurianObject::DurianObject(int32_t len, unsigned char *sval) :
```
Comment (rwblickhan) [inline src/interpreter/vm/DurianObject.cpp:25]: I would prefer to have the `:` for member init on a new line, e.g.
```cpp
DurianObject::DurianObject(int32_t len, unsigned char *sval)
    : type(DurianType::String)
    , otherMember(initVal)
```

At least to me it makes it a little clearer that this is in fact member initialization and also lets you line up the `:` on the first line with the `,` on later lines, which looks nice and symmetric.

File: src/include/DurianString.h [hunk 7]
```diff
@@ -0,0 +1,7 @@
+#pragma once
+
+class DurianString {
+public:
+    int32_t len;
```
Comment (rwblickhan) [inline src/include/DurianString.h]: These should have `m_` names.

File: src/interpreter/vm/VM.cpp [hunk 8]
```diff
@@ -17,21 +17,24 @@ VM::VM(unsigned char *bytecode) :
...
 
 
 int VM::run() {
-    //DurianObject a, b, v;
     while(true) {
         unsigned char opcode = nextBytecode();
         DURIAN_DEBUG_LOG("%x\n", opcode);
         DurianObject a, b;
-        int32_t jumpLen;
+        int32_t jump_len; // Jump length
```
Comment (rwblickhan) [inline src/interpreter/vm/VM.cpp]: Why was this changed from camel case to underscores? I thought we were doing camel case?

File: src/include/DurianObject.h (hunk 1 above)
Comment (rwblickhan) [reply src/include/DurianObject.h:19]: Agreed, I don't really like names like `a.value.sval.val`.

Comment (CoffeeTableEspresso) [reply src/include/DurianObject.h:19]: I don't think it's ideal either but any better ideas? that aren't ridiculously long ideally.

File: src/interpreter/vm/DurianObject.cpp (hunk 6 above)
Comment (CoffeeTableEspresso) [reply src/interpreter/vm/DurianObject.cpp:25]: Oh I personally prefer the other style:

```cpp
DurianObject::DurianObject(int32_t len, unsigned char *sval) :
    type(DurianType::String),
    otherMember(initVal)
```

File: src/interpreter/vm/VM.cpp (hunk 8 above)
Comment (CoffeeTableEspresso) [reply src/interpreter/vm/VM.cpp]: yes we should consistently do camelCase

File: src/interpreter/vm/VM.cpp [hunk 9]
```diff
@@ -174,11 +154,7 @@ int VM::run() {
                 if (a.type == DurianType::String && b.type == DurianType::String) {
                     push(DurianObject(a.value.sval.m_len /* + b.value.sval.m_len */, a.value.sval.p_val /* TODO: Implement concat string */));
                 } else {
-                    std::cout << "TypeError: Invalid operand types for &: "
-                              << a.type
-                              << " and "
-                              << b.type
-                              << "." << std::endl;
+                    typeError("&", a, b);
```
Comment (bobheadxi) [inline src/interpreter/vm/VM.cpp]: sexy ⭐️

Comment (bwdmonkey) [reply src/interpreter/vm/VM.cpp]: ... Go to sleep my man

**Review by rwblickhan — APPROVED** [review]
